File: src/card.rs
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Rank {
    Ace,
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
    Ten,
    Jack,
    Queen,
    King,
}

pub use Rank::*;

#[rustfmt::skip]
pub const RANKS: [Rank; 13] = [
    Ace, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Jack, Queen, King,
];

impl Rank {
    pub fn as_str(self) -> &'static str {
        match self {
            Ace => "A",
            Two => "2",
            Three => "3",
            Four => "4",
            Five => "5",
            Six => "6",
            Seven => "7",
            Eight => "8",
            Nine => "9",
            Ten => "10",
            Jack => "J",
            Queen => "Q",
            King => "K",
        }
    }
}

impl std::fmt::Display for Rank {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}", self.as_str())
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Suit {
    Club,
    Diamond,
    Heart,
    Spade,
}

pub use Suit::*;

#[rustfmt::skip]
pub const SUITS: [Suit; 4] = [
    Club, Diamond, Heart, Spade,
];

impl Suit {
    pub fn as_str(self) -> &'static str {
        match self {
            Club => "♣",
            // Club => "C",
            Diamond => "♦",
            // Diamond => "D",
            Heart => "♥",
            // Heart => "H",
            Spade => "♠",
            // Spade => "S",
        }
    }
}

impl std::fmt::Display for Suit {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}", self.as_str())
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Card {
    rank: Rank,
    suit: Suit,
}

impl Card {
    pub fn new(rank: Rank, suit: Suit) -> Card {
        Card { rank, suit }
    }
}

impl std::fmt::Display for Card {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}{}", self.rank, self.suit)
    }
}

pub fn deck() -> Vec<Card> {
    let mut cards = Vec::with_capacity(52);

    for rank in RANKS {
        for suit in SUITS {
            cards.push(Card::new(rank, suit));
        }
    }

    cards
}

File: packages/card/experiments/card.ts
export enum Rank {
	Ace = 0,
	Two = 1,
	Three = 2,
	Four = 3,
	// ...
}

export namespace Rank {
	export function toCharacter(rank: Rank): string {
		switch (rank) {
			case Rank.Ace:
				return "A";
			case Rank.Two:
				return "2";
			case Rank.Three:
				return "3";
			case Rank.Four:
				return "4";
		}
	}
}

export enum Suit {
	Club = 0,
	Diamond = 1,
	Heart = 2,
	Spade = 3,
}

export namespace Suit {
	export function toCharacter(suit: Suit): string {
		switch (suit) {
			case Suit.Club:
				return "C";
			case Suit.Diamond:
				return "D";
			case Suit.Heart:
				return "H";
			case Suit.Spade:
				return "S";
		}
	}
}

export interface Card {
	rank: Rank;
	suit: Suit;
}

export namespace Card {
	export function create(rank: Rank, suit: Suit): Card {
		return { rank, suit };
	}

	export function toIdentifier(card: Card): string {
		return `${Rank.toCharacter(card.rank)}${Suit.toCharacter(card.suit)}`;
	}
}

const card = Card.create(Rank.Ace, Suit.Club);
const identifier = Card.toIdentifier(card);
